test/python/test_ttir_ops.py

```diff
@@ -0,0 +1,112 @@
+# SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+
+# RUN: pytest -svv %s | FileCheck %s
+
+from test_infra.ttir import Add, Multiply, Exp
+from test_infra.ttir_builder import TTIRBuilder
+from ttmlir.ir import *
+
+import torch
+import pytest
+
+# ----- Fixtures -----
+
+
+@pytest.fixture
+def context() -> Context:
+    return Context()
+
+
+@pytest.fixture
+def location(context: Context) -> Location:
+    return Location.unknown(context)
+
+
+@pytest.fixture
+def builder(context: Context, location: Location) -> TTIRBuilder:
+    return TTIRBuilder(context, location)
+
+
+@pytest.fixture
+def add(context: Context, location: Location) -> Add:
+    return Add(context, location)
+
+
+@pytest.fixture
+def multiply(context: Context, location: Location) -> Multiply:
+    return Multiply(context, location)
+
+
+@pytest.fixture
+def exp(context: Context, location: Location) -> Exp:
+    return Exp(context, location)
+
+
+# ----- Tests -----
+
+
+def test_add(builder: TTIRBuilder, add: Add):
+    in0_shape = (32, 32)
+    in1_shape = (32, 32)
+    out_shape = (32, 32)
+
+    f = builder.emit_mlir_function(add, [in0_shape, in1_shape], [out_shape])
```

Comment:
Ok so I was just rereading the issue and I think we should optimize for test simplicity.  Ideally the whole test is something like:

```python
def test_add(builder: TTIRBuilder):
    in0 = builder.empty(32, 32)
    in1 = builder.empty(32, 32)
    out = builder.add(in0, in1)
```

So I'm not necessarily suggesting that we get rid of the ABC hierarchy, but just to illustrate the idea more succinctly, inside of builder we have all of the top level functions:

```python
class TTIRBuilder:
    ...

    def empty(*shape, rand_fn=torch.randn):
        with self.ctx, self.loc:
            result = tensor.EmptyOp(...)
            self.goldens[result] = rand_fn(shape, seed = self.get_seed())
            return result

    def add(self, a, b):
        with self.ctx, self.loc:
            output = builder.empty(a.shape)
            result = ttir.AddOp(
                [output.type],
                [a, b],
                [output],
                self._wrap_operand_constraints(
                    3, operand_constraints
                ),
            )
            self.goldens[result] = torch.add(self.goldens[a], self.goldens[b])
            return result
```

This enables the tests to string together arbitrarily complex graphs, `builder.goldens` tracks all intermediates and contains everything needed to build the MLIR, convert to flatbuffer, and embed the goldens in the flatbuffer.

Let me know what you think.

Reply:
The [example](https://github.com/tenstorrent/tt-mlir/pull/854#discussion_r1791067535) you provided looks neat and clean to me, but I just was not able to implement it this simply.

Main reason is that this

```python
class TTIRBuilder:
    ...
    def add(self, in0: Operand, in1: Operand) -> OpView:
        """Convenience wrapper constructing `ttir.AddOp`."""
        with self.ctx, self.loc:
            output = self.empty(self.get_shape(in0))

            return ttir.AddOp(
                [self.__get_type(output)],
                [in0, in1],
                [output],
                self.__get_operand_constraint_attr(3),
            )


def test_add(builder: TTIRBuilder):
    in0 = builder.empty((32, 32))
    in1 = builder.empty((32, 32))
    add = builder.add(in0, in1)
    print(add)


if __name__ == "__main__":
    ctx = Context()
    loc = Location.unknown(ctx)
    builder = TTIRBuilder(ctx, loc)
    test_add(builder)
```

will end up with

```bash
error: 'tensor.empty' op operation destroyed but still has uses
LLVM ERROR: operation destroyed but still has uses
Aborted (core dumped)
```

in any case I tried. Seems to be a dangling reference to output `tensor.EmptyOp`. Even if I do 

```python
class TTIRBuilder:
    ...
    def add(self, in0: Operand, in1: Operand, output) -> OpView:
        """Convenience wrapper constructing `ttir.AddOp`."""
        with self.ctx, self.loc:
            return ttir.AddOp(
                [self.__get_type(output)],
                [in0, in1],
                [output],
                self.__get_operand_constraint_attr(3),
            )


def test_add(builder: TTIRBuilder):
    in0 = builder.empty((32, 32))
    in1 = builder.empty((32, 32))
    out = builder.empty((32, 32))
    add = builder.add(in0, in1, out)
    print(add)
```

to make sure `out` doesn't go out of scope, still the same thing happens. 

However, if I wrap the graph in a MLIR function, tie inputs properly and return the output (@vprajapati-tt helped me realize this), it works. Thus I changed approach, and used what `tensor` dialect used in its [examples](https://github.com/llvm/llvm-project/blob/main/mlir/test/python/dialects/tensor.py).

I think your [example](https://github.com/tenstorrent/tt-mlir/pull/854#discussion_r1792537840) with `builder.finish` was suggesting something like that.